File: frontend/lib/api/imports.ts
import { apiClient } from '@/lib/api/client';
import type {
  BulkApplyPayload,
  BulkApplyResponse,
  BulkClustersResponse,
  ImportCommitResponse,
  ImportMappingPayload,
  ImportPreviewResponse,
  ImportPreviewRow,
  ImportQueueBulkClustersResponse,
  ImportQueueCommitResponse,
  ImportQueuePreviewResponse,
  ImportQueueStartAllResponse,
  ImportSessionResponse,
  ImportSessionListItem,
  ImportReviewQueueResponse,
  ImportRowUpdatePayload,
  ImportRowUpdateResponse,
  ImportUploadResponse,
  ModerationStatusResponse,
  ParkedQueueResponse,
} from '@/types/import';

export function uploadImportFile(payload: {
  file: File;
  delimiter: string;
  // Этап 0.5: bypass duplicate-file detection. Set after the user picks
  // [Перезаписать] (active duplicate) or [Загрузить как новую] (committed
  // duplicate) in DuplicateStatementModal. Backend creates a new parallel
  // session; the existing one is preserved.
  forceNew?: boolean;
}) {
  const formData = new FormData();
  formData.set('file', payload.file);
  formData.set('delimiter', payload.delimiter);

  // Backend reads `force_new` as a query parameter (FastAPI query param
  // semantics — easier than a multipart string field for a boolean flag).
  const url = payload.forceNew
    ? '/imports/upload?force_new=true'
    : '/imports/upload';

  return apiClient<ImportUploadResponse>(url, {
    method: 'POST',
    body: formData,
  });
}

export function previewImport(sessionId: number, payload: ImportMappingPayload) {
  return apiClient<ImportPreviewResponse>(`/imports/${sessionId}/preview`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export function getImportPreview(sessionId: number) {
  return apiClient<ImportPreviewResponse>(`/imports/${sessionId}/preview`);
}

export function commitImport(sessionId: number, importReadyOnly = true) {
  return apiClient<ImportCommitResponse>(`/imports/${sessionId}/commit`, {
    method: 'POST',
    body: JSON.stringify({ import_ready_only: importReadyOnly }),
  });
}

export function getImportSessions() {
  return apiClient<{ sessions: ImportSessionListItem[]; total: number }>('/imports/sessions');
}

// ──────────────────────────────────────────────────────────────────
// Unified queue (cross-session moderation, v1.23)
// ──────────────────────────────────────────────────────────────────

export function getImportQueuePreview() {
  return apiClient<ImportQueuePreviewResponse>('/imports/queue/preview');
}

export function getImportQueueBulkClusters() {
  return apiClient<ImportQueueBulkClustersResponse>('/imports/queue/bulk-clusters');
}

export function commitImportQueueConfirmed() {
  return apiClient<ImportQueueCommitResponse>('/imports/queue/commit-confirmed', {
    method: 'POST',
  });
}

export function startImportQueueAll() {
  return apiClient<ImportQueueStartAllResponse>('/imports/queue/start-all', {
    method: 'POST',
  });
}

export function getImportSession(sessionId: number) {
  return apiClient<ImportSessionResponse>(`/imports/${sessionId}`);
}

export function deleteImportSession(sessionId: number) {
  return apiClient<void>(`/imports/${sessionId}`, {
    method: 'DELETE',
  });
}


export function getImportReviewQueue() {
  return apiClient<ImportReviewQueueResponse>('/imports/review-queue');
}


export function assignSessionAccount(sessionId: number, accountId: number) {
  return apiClient<ImportSessionListItem>(`/imports/${sessionId}/account`, {
    method: 'PATCH',
    body: JSON.stringify({ account_id: accountId }),
  });
}

export function updateImportRow(rowId: number, payload: ImportRowUpdatePayload) {
  return apiClient<ImportRowUpdateResponse>(`/imports/rows/${rowId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  });
}

// LLM moderation removed from the import pipeline (decision 2026-05-03).
// Stubs return immediately without hitting the backend (which now returns 410).
// Kept so any straggler imports compile while we delete call sites.
export async function getModerationStatus(_sessionId: number): Promise<ModerationStatusResponse> {
  return {
    session_id: _sessionId,
    status: 'disabled',
    total_clusters: 0,
    processed_clusters: 0,
    started_at: null,
    finished_at: null,
    error: null,
  } as unknown as ModerationStatusResponse;
}

export async function startModeration(sessionId: number) {
  return { session_id: sessionId, status: 'disabled' };
}

export function getBulkClusters(sessionId: number) {
  return apiClient<BulkClustersResponse>(`/imports/${sessionId}/clusters`);
}

export function bulkApplyCluster(sessionId: number, payload: BulkApplyPayload) {
  return apiClient<BulkApplyResponse>(`/imports/${sessionId}/clusters/bulk-apply`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export type AttachRowToClusterResponse = {
  row_id: number;
  transaction_id: number | null;
  target_fingerprint: string | null;
  counterparty_id: number | null;
  alias_created: boolean;
  binding_created: boolean;
  source_fingerprint: string | null;
  summary: Record<string, number>;
};

// Legacy: attach to a fingerprint cluster (creates FingerprintAlias).
export function attachRowToCluster(
  sessionId: number,
  rowId: number,
  targetFingerprint: string,
) {
  return apiClient<AttachRowToClusterResponse>(
    `/imports/${sessionId}/rows/${rowId}/attach-to-cluster`,
    {
      method: 'POST',
      body: JSON.stringify({ target_fingerprint: targetFingerprint }),
    },
  );
}

// Preferred Phase 3 path: attach to a counterparty. Creates a
// CounterpartyFingerprint binding so future imports of the same skeleton
// group under the counterparty automatically.
export function attachRowToCounterparty(
  sessionId: number,
  rowId: number,
  counterpartyId: number,
) {
  return apiClient<AttachRowToClusterResponse>(
    `/imports/${sessionId}/rows/${rowId}/attach-to-cluster`,
    {
      method: 'POST',
      body: JSON.stringify({ counterparty_id: counterpartyId }),
    },
  );
}

// Brand confirm/reject (Brand registry Ph6 + Ph8).
export type BrandConfirmResponse = {
  row_id: number;
  brand_id: number;
  brand_slug: string;
  brand_canonical_name: string;
  counterparty_id: number | null;
  counterparty_name: string | null;
  category_id: number | null;
  category_name: string | null;
  propagated_count: number;
  was_override: boolean;
};

export type BrandRejectResponse = {
  row_id: number;
  rejected_brand_id: number;
};

export function confirmRowBrand(
  rowId: number,
  brandId: number,
  categoryId?: number | null,
) {
  // Ph8: optional categoryId overrides the brand's default hint and saves
  // a per-user override so future imports of this brand resolve to the
  // chosen category.
  const body: Record<string, unknown> = { brand_id: brandId };
  if (categoryId != null) body.category_id = categoryId;
  return apiClient<BrandConfirmResponse>(
    `/imports/rows/${rowId}/confirm-brand`,
    { method: 'POST', body: JSON.stringify(body) },
  );
}

export function rejectRowBrand(rowId: number) {
  return apiClient<BrandRejectResponse>(
    `/imports/rows/${rowId}/reject-brand`,
    { method: 'POST', body: JSON.stringify({}) },
  );
}

// Ph8: bulk-apply category for an entire brand (post-confirmation editing).
export type ApplyBrandCategoryResponse = {
  brand_id: number;
  brand_canonical_name: string;
  category_id: number;
  category_name: string;
  rows_updated: number;
  override_id: number;
};

export function applyBrandCategory(brandId: number, categoryId: number) {
  return apiClient<ApplyBrandCategoryResponse>(
    `/brands/${brandId}/apply-category`,
    {
      method: 'POST',
      body: JSON.stringify({ category_id: categoryId }),
    },
  );
}

export function parkImportRow(rowId: number) {
  return apiClient<{ session_id: number; row_id: number; status: string; summary: Record<string, number> }>(
    `/imports/rows/${rowId}/park`,
    { method: 'POST', body: JSON.stringify({}) },
  );
}

export function unparkImportRow(rowId: number) {
  return apiClient<{ session_id: number; row_id: number; status: string; summary: Record<string, number> }>(
    `/imports/rows/${rowId}/unpark`,
    { method: 'POST', body: JSON.stringify({}) },
  );
}

export function excludeImportRow(rowId: number) {
  return apiClient<{ session_id: number; row_id: number; status: string; summary: Record<string, number> }>(
    `/imports/rows/${rowId}/exclude`,
    { method: 'POST', body: JSON.stringify({}) },
  );
}

export function unexcludeImportRow(rowId: number) {
  return apiClient<{ session_id: number; row_id: number; status: string; summary: Record<string, number> }>(
    `/imports/rows/${rowId}/unexclude`,
    { method: 'POST', body: JSON.stringify({}) },
  );
}

export function unpairImportRow(rowId: number) {
  return apiClient<{ session_id: number; row_id: number; status: string; summary: Record<string, number> }>(
    `/imports/rows/${rowId}/unpair`,
    { method: 'POST', body: JSON.stringify({}) },
  );
}

export function detachImportRowFromCluster(rowId: number) {
  return apiClient<{ session_id: number; row_id: number; status: string; summary: Record<string, number> }>(
    `/imports/rows/${rowId}/detach-from-cluster`,
    { method: 'POST', body: JSON.stringify({}) },
  );
}

export function getParkedQueue() {
  return apiClient<ParkedQueueResponse>('/imports/parked-queue');
}

export function rematchTransfers() {
  return apiClient<{ status: string }>('/imports/rematch-transfers', { method: 'POST' });
}

File: app/api/v1/imports.py
from fastapi import APIRouter, Depends, File, HTTPException, Request, UploadFile, status
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.api.v1._upload_helpers import validate_and_read_upload
from app.core.config import settings
from app.core.keys import user_or_ip_key
from app.core.rate_limit import limiter
from app.models.user import User
from app.schemas.imports import (
    AttachRowToClusterRequest,
    AttachRowToClusterResponse,
    BrandConfirmRequest,
    BrandConfirmResponse,
    BrandRejectResponse,
    BulkApplyRequest,
    BulkApplyResponse,
    BulkClustersResponse,
    ImportCommitRequest,
    ImportCommitResponse,
    ImportMappingRequest,
    ImportPreviewResponse,
    ImportQueueBulkClustersResponse,
    ImportQueuePreviewResponse,
    ImportSessionListResponse,
    ImportReviewQueueResponse,
    ImportRowLabelRequest,
    ImportRowLabelResponse,
    ImportRowUpdateRequest,
    ImportRowUpdateResponse,
    ImportSessionResponse,
    ImportUploadResponse,
)
from app.services.import_service import (
    BankUnsupportedError,
    ImportNotFoundError,
    ImportService,
    ImportValidationError,
)
from app.services.upload_validator import UnsupportedUploadTypeError, UploadTooLargeError

router = APIRouter(prefix="/imports", tags=["Imports"])


@router.get("/review-queue", response_model=ImportReviewQueueResponse)
def get_import_review_queue(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    return service.list_review_queue(user_id=current_user.id)


@router.get("/queue/preview", response_model=ImportQueuePreviewResponse)
def get_import_queue_preview(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Unified moderation queue (v1.23) — all preview-ready rows from
    every active session of the user, with per-row source metadata.

    The session abstraction is hidden from the user: this endpoint is the
    single source of truth for the moderator UI in queue mode. Only
    sessions with `status='preview_ready'` AND assigned `account_id` are
    admitted; sessions still queued / parsing / awaiting account are
    excluded silently.
    """
    service = ImportService(db)
    return service.get_queue_preview(user_id=current_user.id)


@router.get("/queue/bulk-clusters", response_model=ImportQueueBulkClustersResponse)
def get_import_queue_bulk_clusters(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Cross-session bulk clusters (v1.23). Aggregates fingerprint clusters
    from every preview-ready session of the user; brand and counterparty
    groups span sessions naturally.

    Brand-level threshold (≥2 fingerprints, ≥2 rows) now operates over
    the union of clusters — «Магнит ×1 в Сбере + Магнит ×1 в Т-Банке»
    rolls into one BrandCluster, which is the whole point of cross-session
    moderation.
    """
    service = ImportService(db)
    return service.get_queue_bulk_clusters(user_id=current_user.id)


@router.get("/sessions", response_model=ImportSessionListResponse)
def list_import_sessions(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Список незавершённых сессий импорта текущего пользователя."""
    service = ImportService(db)
    return service.list_active_sessions(user_id=current_user.id)


@router.post("/upload", status_code=status.HTTP_201_CREATED)
@limiter.limit(settings.RATE_LIMIT_UPLOAD, key_func=user_or_ip_key)
async def upload_import_file(
    request: Request,
    file: UploadFile = File(...),
    delimiter: str = ",",
    force_new: bool = False,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    # response_model intentionally omitted — JSONResponse from validation
    # rejects bypasses Pydantic serialization, and ImportUploadResponse on the
    # success path is documented at the function level instead.
    #
    # `force_new=true` (Этап 0.5) bypasses file_hash deduplication and creates
    # a new parallel session even if a duplicate exists. The frontend sends
    # this after the user picks "Перезаписать" / "Загрузить как новую" in
    # the duplicate-detection modal.
    try:
        raw_bytes, _detected = await validate_and_read_upload(file)
    except UploadTooLargeError as exc:
        return JSONResponse(status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, content=exc.to_payload())
    except UnsupportedUploadTypeError as exc:
        return JSONResponse(status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE, content=exc.to_payload())

    service = ImportService(db)
    try:
        return service.upload_source(
            user_id=current_user.id,
            filename=file.filename or "import_file",
            raw_bytes=raw_bytes,
            delimiter=delimiter,
            force_new=force_new,
        )
    except BankUnsupportedError as exc:
        # 415 (not 400) — same family as upload_validator's
        # `extension_content_mismatch` / `unsupported_upload_type`. The frontend
        # branches on `code`, so the status code is informational; we keep it
        # within the "uploaded media is wrong shape" semantic group.
        db.rollback()
        return JSONResponse(
            status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            content={
                "code": "bank_unsupported",
                "bank_id": exc.bank_id,
                "bank_name": exc.bank_name,
                "extractor_status": exc.extractor_status,
                "detail": str(exc),
            },
        )
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc





# LLM moderation endpoints — DISABLED (decision 2026-05-03).
# The async LLM moderator created more friction than value: token cost without
# meaningful confidence lift over the deterministic rule + bank_mechanics layers.
# Endpoints return 410 Gone to surface stale frontend callers; full removal of
# the underlying services / Celery task is tracked as a follow-up.
@router.get("/{session_id}/moderation-status")
@router.post("/{session_id}/moderate")
@router.get("/moderation-metrics")
def _moderation_disabled():
    raise HTTPException(
        status_code=status.HTTP_410_GONE,
        detail="LLM moderation has been removed from the import pipeline.",
    )


@router.post("/rematch-transfers")
def rematch_transfers(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Re-run the global transfer matcher across all user sessions.

    Use when sessions were uploaded/removed and cross-session pairs may have
    shifted. Cheap — pure DB work, no LLM. Goes through the debounced matcher
    so manual + automatic triggers coalesce into a single run.
    """
    from app.jobs.transfer_matcher_debounced import schedule_transfer_match

    schedule_transfer_match(current_user.id)
    return {"status": "queued"}


@router.get("/parked-queue")
def get_parked_queue(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    return service.list_parked_queue(user_id=current_user.id)


@router.post("/rows/{row_id}/park")
def park_import_row(
    row_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.park_row(user_id=current_user.id, row_id=row_id)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/rows/{row_id}/unpark")
def unpark_import_row(
    row_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.unpark_row(user_id=current_user.id, row_id=row_id)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/rows/{row_id}/exclude")
def exclude_import_row(
    row_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.exclude_row(user_id=current_user.id, row_id=row_id)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/rows/{row_id}/unexclude")
def unexclude_import_row(
    row_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.unexclude_row(user_id=current_user.id, row_id=row_id)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/rows/{row_id}/unpair")
def unpair_import_row(
    row_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.unpair_row(user_id=current_user.id, row_id=row_id)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/rows/{row_id}/detach-from-cluster")
def detach_row_from_cluster(
    row_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.detach_row_from_cluster(user_id=current_user.id, row_id=row_id)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/rows/{row_id}/label", response_model=ImportRowLabelResponse)
def set_import_row_label(
    row_id: int,
    payload: ImportRowLabelRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.set_row_label(user_id=current_user.id, row_id=row_id, user_label=payload.user_label)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.patch("/rows/{row_id}", response_model=ImportRowUpdateResponse)
def update_import_row(
    row_id: int,
    payload: ImportRowUpdateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.update_row(user_id=current_user.id, row_id=row_id, payload=payload)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/rows/{row_id}/confirm-brand", response_model=BrandConfirmResponse)
def confirm_row_brand(
    row_id: int,
    payload: BrandConfirmRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Confirm the brand for a row («Это Пятёрочка? [Да]»).

    When `brand_id` matches the resolver's prediction, the same brand is
    propagated to every other row of this session that resolved to it —
    the prompt vanishes across the cluster in one click. When it differs
    (override), only the current row is stamped; the predicted pattern
    gets a rejection signal.
    """
    from app.services.brand_confirm_service import BrandConfirmError, BrandConfirmService

    service = BrandConfirmService(db)
    try:
        return service.confirm_brand_for_row(
            user_id=current_user.id,
            row_id=row_id,
            brand_id=payload.brand_id,
            category_id=payload.category_id,
        )
    except BrandConfirmError as exc:
        db.rollback()
        # 400 covers all rejection causes — row not found, session committed,
        # row committed, brand missing/inaccessible. The message is in Russian
        # and surfaced verbatim to the moderator UI.
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/rows/{row_id}/reject-brand", response_model=BrandRejectResponse)
def reject_row_brand(
    row_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Reject the predicted brand for a row («Не этот бренд»).

    Stamps `user_rejected_brand_id` so the resolver doesn't re-suggest the
    same brand on next preview-read. Increments the predicted pattern's
    rejection counter; pattern auto-deactivates after enough rejections.
    No propagation — rejection is row-local.
    """
    from app.services.brand_confirm_service import BrandConfirmError, BrandConfirmService

    service = BrandConfirmService(db)
    try:
        return service.reject_brand_for_row(
            user_id=current_user.id, row_id=row_id,
        )
    except BrandConfirmError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.get("/{session_id}/clusters", response_model=BulkClustersResponse)
def get_bulk_clusters(
    session_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return bulk-eligible clusters (И-08 Этап 2).

    Fingerprint clusters of size ≥5, plus brand-level groups that aggregate
    ≥2 fingerprints into a single row-count ≥5 block. Transfer-secondary
    rows are pre-filtered so they never appear in the bulk UI — they're
    auto-created via their transfer pair on commit.
    """
    service = ImportService(db)
    try:
        return service.get_bulk_clusters(user_id=current_user.id, session_id=session_id)
    except ImportNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.post(
    "/{session_id}/rows/{row_id}/attach-to-cluster",
    response_model=AttachRowToClusterResponse,
)
def attach_row_to_cluster(
    session_id: int,
    row_id: int,
    payload: AttachRowToClusterRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Move a "needs attention" row into an existing cluster.

    Creates a user-scoped FingerprintAlias so future imports with the same
    source pattern land in the target cluster automatically (Level 3). The
    row is committed atomically using the target cluster's suggested category.
    """
    service = ImportService(db)
    try:
        return service.attach_row_to_cluster(
            user_id=current_user.id,
            session_id=session_id,
            row_id=row_id,
            target_fingerprint=payload.target_fingerprint,
            counterparty_id=payload.counterparty_id,
        )
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/{session_id}/clusters/bulk-apply", response_model=BulkApplyResponse)
def bulk_apply_cluster(
    session_id: int,
    payload: BulkApplyRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Apply a moderator action to all included rows of a cluster.

    One click, N confirmations (see project_bulk_clusters.md). Each row is
    confirmed via the same path as a manual single-row update. Rules are
    upserted per unique (fingerprint, category) combination with
    `confirms_delta = N` — a single bulk action activates AND generalizes
    the rule in one transition.
    """
    service = ImportService(db)
    try:
        return service.bulk_apply_cluster(
            user_id=current_user.id, session_id=session_id, payload=payload,
        )
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.get("/{session_id}", response_model=ImportSessionResponse)
def get_import_session(
    session_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.get_session(user_id=current_user.id, session_id=session_id)
    except ImportNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.patch("/{session_id}/account")
def assign_session_account(
    session_id: int,
    payload: dict,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Assign (or re-assign) an account to an import session and re-run transfer matching."""
    service = ImportService(db)
    session = service.import_repo.get_session(session_id=session_id, user_id=current_user.id)
    if session is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Сессия не найдена.")
    account_id = payload.get("account_id")
    if not account_id:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="account_id обязателен.")
    service.import_repo.update_session(session, account_id=int(account_id))
    db.commit()
    db.refresh(session)

    # If the session was uploaded without a detected account (very common when
    # the user had no existing statements for this bank yet), auto-preview was
    # skipped on upload. Now that we have an account, fire it — so the bulk
    # flow "upload N statements → assign each to an account" matches the
    # queue-era UX where rows get extracted and matched without manual clicks.
    if session.status == "analyzed":
        mapping = session.mapping_json or {}
        field_mapping = mapping.get("field_mapping") or {}
        if field_mapping.get("date") and field_mapping.get("amount"):
            try:
                from app.jobs.auto_preview_import_session import auto_preview_import_session
                auto_preview_import_session.delay(session.id)
            except Exception:
                pass

    # Trigger the debounced global transfer matcher so existing previewed
    # sessions can now match against this one.
    try:
        from app.jobs.transfer_matcher_debounced import schedule_transfer_match
        schedule_transfer_match(current_user.id)
    except Exception:
        pass
    return {
        "id": session.id,
        "filename": session.filename,
        "source_type": session.source_type,
        "status": session.status,
        "account_id": session.account_id,
        "created_at": session.created_at,
        "updated_at": session.updated_at,
    }


@router.delete("/{session_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_import_session(
    session_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        service.delete_session(user_id=current_user.id, session_id=session_id)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.get("/{session_id}/preview", response_model=ImportPreviewResponse)
def get_import_preview(
    session_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.get_existing_preview(user_id=current_user.id, session_id=session_id)
    except ImportNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.post("/{session_id}/preview", response_model=ImportPreviewResponse)
def preview_import(
    session_id: int,
    payload: ImportMappingRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.build_preview(user_id=current_user.id, session_id=session_id, payload=payload)
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/{session_id}/commit", response_model=ImportCommitResponse)
def commit_import(
    session_id: int,
    payload: ImportCommitRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ImportService(db)
    try:
        return service.commit_import(
            user_id=current_user.id,
            session_id=session_id,
            import_ready_only=payload.import_ready_only,
        )
    except ImportNotFoundError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ImportValidationError as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
